File: pkg-py/src/biogate/_pattern.py
"""Pattern mode: offline, deterministic checks against a source regex.

The regexes live in the shared spec and use ASCII classes so that R (PCRE) and
Python agree element for element.
"""

from __future__ import annotations

import functools
import re

from ._registry import Source

_LOCAL_DIGITS = re.compile("[0-9]+")

# ENS, a greedy species code, one feature letter, then 11 digits. The greedy
# [A-Z]* plus the single [EFGPT] yields "" for ENSG... and "MUS" for ENSMUSG...
_ENSEMBL_ID = re.compile("^ENS([A-Z]*)[EFGPT][0-9]{11}$")


@functools.lru_cache(maxsize=256)
def _compiled(pattern: str) -> re.Pattern:
    return re.compile(pattern)


def matches(pattern: str, s: str) -> bool:
    """Whether ``s`` matches ``pattern`` in full."""
    return _compiled(pattern).fullmatch(s) is not None


def _suggest(source: Source, s: str) -> str | None:
    curie = source.curie
    if curie:
        prefix = curie["prefix"]
        pad_to = curie.get("pad_to")
        head, sep, local = s.partition(":")
        if not sep:
            head, local = prefix, s
        if head.upper() != prefix.upper():
            return None
        if pad_to and _LOCAL_DIGITS.fullmatch(local):
            local = local.zfill(int(pad_to))
        candidate = f"{prefix}:{local}"
        if candidate != s and matches(source.pattern, candidate):
            return candidate
        return None

    norm = source.normalize
    if norm and norm.get("case") in ("upper", "lower"):
        candidate = s.upper() if norm["case"] == "upper" else s.lower()
        if candidate != s and matches(source.pattern, candidate):
            return candidate
    return None


def _ensembl_id_prefix(ident: str) -> str | None:
    """Return the species code between ENS and the feature letter.

    Returns "" for a human id (ENSG...), the code for other species (for
    example "MUS" for ENSMUSG...), or None when ``ident`` is not an Ensembl
    stable id.
    """
    m = _ENSEMBL_ID.match(ident)
    if m is None:
        return None
    return m.group(1)


def _ensembl_species_prefix(species_block: dict, species) -> str | None:
    """Return the expected id prefix for ``species`` from the species map.

    ``species`` may be a name like "homo_sapiens" or a taxon id like 9606.
    Returns the mapped prefix (which may be "") or None when the species is
    not in the map.
    """
    for entry in species_block.get("map", []):
        if str(entry.get("name")) == str(species) or entry.get("taxon") == species:
            return entry.get("prefix")
    return None


def _species_ok(source: Source, ident: str, species) -> bool:
    """Whether ``ident`` is consistent with the requested ``species``.

    Lenient: a missing species, a source without a species scheme we know, or
    a species outside the map all pass. A non-Ensembl-shaped id also passes;
    the base pattern decides shape.
    """
    if species is None:
        return True
    block = source.species
    if not block or block.get("scheme") != "ensembl_prefix":
        return True
    expected = _ensembl_species_prefix(block, species)
    if expected is None:
        return True
    id_prefix = _ensembl_id_prefix(ident)
    if id_prefix is None:
        return True
    return id_prefix == expected


def check_one(
    source: Source, s: str, species=None
) -> tuple[bool, str | None, str | None]:
    """Return ``(valid, normalized, suggestion)`` for a single input."""
    if matches(source.pattern, s):
        if _species_ok(source, s, species):
            return True, s, None
        return False, None, None
    candidate = _suggest(source, s)
    if candidate is not None and _species_ok(source, candidate, species):
        return False, None, candidate
    return False, None, None

File: pkg-js/src/pattern.ts
// Pattern mode: offline, deterministic checks against a source regex, ported
// from pkg-py/src/biobouncer/_pattern.py. The regexes use ASCII classes so R
// (PCRE), Python (re), and JavaScript (RegExp) agree element for element.

import type { SourceSpec } from "./registry";

const compiled = new Map<string, RegExp>();

/** Whether `s` matches `pattern` in full. */
export function matches(pattern: string, s: string): boolean {
  let re = compiled.get(pattern);
  if (re === undefined) {
    // Wrap in a non-capturing group so the anchors bind to the whole pattern,
    // not just one branch of a top-level alternation. No flags: JS `$` without
    // `m` matches only true end-of-input, equal to Python's re.fullmatch.
    re = new RegExp(`^(?:${pattern})$`);
    compiled.set(pattern, re);
  }
  return re.test(s);
}

const DIGITS = /^[0-9]+$/;

/** Split on the first occurrence of `sep`: [head, found, tail]. */
export function partitionFirst(s: string, sep: string): [string, boolean, string] {
  const i = s.indexOf(sep);
  if (i < 0) return [s, false, ""];
  return [s.slice(0, i), true, s.slice(i + sep.length)];
}

/** A repair for `s`, or null: a wrong-case or unpadded form that maps to a valid id. */
export function suggest(spec: SourceSpec, s: string): string | null {
  const curie = spec.curie;
  if (curie) {
    const prefix = curie.prefix;
    const padTo = curie.pad_to;
    let [head, sep, local] = partitionFirst(s, ":");
    if (!sep) {
      head = prefix;
      local = s;
    }
    if (head.toUpperCase() !== prefix.toUpperCase()) return null;
    if (padTo && DIGITS.test(local)) {
      local = local.padStart(padTo, "0");
    }
    const candidate = `${prefix}:${local}`;
    if (candidate !== s && matches(spec.pattern, candidate)) return candidate;
    return null;
  }

  const norm = spec.normalize;
  if (norm && (norm.case === "upper" || norm.case === "lower")) {
    const candidate = norm.case === "upper" ? s.toUpperCase() : s.toLowerCase();
    if (candidate !== s && matches(spec.pattern, candidate)) return candidate;
  }
  return null;
}

// ENS, a greedy species code, one feature letter, then 11 digits. The greedy
// [A-Z]* plus the single [EFGPT] yields "" for ENSG... and "MUS" for ENSMUSG...
const ENSEMBL_ID = /^ENS([A-Z]*)[EFGPT][0-9]{11}$/;

function ensemblIdPrefix(ident: string): string | null {
  const m = ENSEMBL_ID.exec(ident);
  if (m === null) return null;
  return m[1] ?? "";
}

function ensemblSpeciesPrefix(
  block: NonNullable<SourceSpec["species"]>,
  species: string | number,
): string | null {
  for (const entry of block.map) {
    // Name match stringifies both sides; taxon match fires only for a numeric
    // species, so 9606 matches taxon 9606 but "9606" does not.
    if (
      String(entry.name) === String(species) ||
      (typeof species === "number" && entry.taxon === species)
    ) {
      return entry.prefix ?? null;
    }
  }
  return null;
}

function speciesOk(
  spec: SourceSpec,
  ident: string,
  species: string | number | null,
): boolean {
  if (species === null) return true;
  const block = spec.species;
  if (!block) return true;
  if (block.scheme !== "ensembl_prefix") return true;
  const expected = ensemblSpeciesPrefix(block, species);
  if (expected === null) return true;
  const idPrefix = ensemblIdPrefix(ident);
  if (idPrefix === null) return true;
  return idPrefix === expected;
}

export interface PatternVerdict {
  valid: boolean;
  normalized: string | null;
  suggestion: string | null;
}

/** Return the pattern-mode verdict for a single input. */
export function checkOne(
  spec: SourceSpec,
  s: string,
  species: string | number | null = null,
): PatternVerdict {
  if (matches(spec.pattern, s)) {
    if (speciesOk(spec, s, species)) {
      return { valid: true, normalized: s, suggestion: null };
    }
    return { valid: false, normalized: null, suggestion: null };
  }
  const candidate = suggest(spec, s);
  if (candidate !== null && speciesOk(spec, candidate, species)) {
    return { valid: false, normalized: null, suggestion: candidate };
  }
  return { valid: false, normalized: null, suggestion: null };
}
